### ISSUE TITLE
Updates to data in ConfigMap and Secrets may never be reflected in the application

### ISSUE THREAD

**Issue by PhilipGough** [prometheus-operator/prometheus-operator#4467]
**What happened?**

In parts of the codebase we use `EnvVarSource` from ConfigMaps and Secrets. [For example with Thanos tracing config](https://github.com/prometheus-operator/prometheus-operator/blob/9c0db5656f04e005de6a0413fd8eb8f11ec99757/pkg/thanos/statefulset.go#L285).

From what I can tell, even after the operator reconciles, such components will not be updated with the changes and such an update would not come into effect until the Pods were restarted for whatever reason.

I also believe we may have a similar situation anywhere we use `ConfigMapKeySelector ` or `SecretKeySelector` and mount them as volume, since although the data in the volume would be updated, in many cases, the contents are read by the component on startup, which again would require a restart to reflect the updated state

**Did you expect to see something different?**

I would expect that changes to underlying objects are reflected in the Pods that use them or we provide some documentation on the actions required by the end user otherwise.

**How to reproduce it (as minimally and precisely as possible)**:

**Environment**

* Prometheus Operator version:

    `Insert image tag or Git SHA here`
    
    

* Kubernetes version information:

    `kubectl version`
    

* Kubernetes cluster kind:

    insert how you created your cluster: kops, bootkube, etc.

* Manifests:

```
insert manifests relevant to the issue
```

* Prometheus Operator Logs:

```
insert Prometheus Operator logs relevant to the issue here
```

**Anything else we need to know?**:

**Comment by simonpasquier** [comment]
I agree with your analysis. For things like the Thanos tracing config, the operator shouldn't use environment variables and inject the configuration directly in the arguments list instead:

```yaml
  args:
  - ...
  - |-
    --tracing.config=type: JAEGER
          config:
            service_name: ...
  - ...
```

**Comment by paulfantom** [comment]
As remediation, we can compute and add checksums for objects included via `ConfigMapKeySelector`, `SecretKeySelector`, and `EnvVarSource` as pod annotations to trigger reconciliation due to annotation change (example: https://github.com/thaum-xyz/ankhmorpork/blob/master/apps/homer/manifests/deployment.yaml#L20-L21).

**Comment by simonpasquier** [comment]
@paulfantom will this trigger a recreation of the pods?

**Comment by paulfantom** [comment]
@simonpasquier yes, that's exactly what will happen

**Comment by rexagod** [comment]
`EnvVarSource` cases addressed in #6049.

**Comment by mviswanathsai** [comment]
Is there any work currently being done on this?

**Comment by simonpasquier** [comment]
I don't think that anybody's actively working on it. My understanding (at least for Thanos Ruler) is that the operator [doesn't pass arguments through env vars](https://github.com/prometheus-operator/prometheus-operator/blob/35340a6c4a97a9e6899f3d0319f6c47f781c2062/pkg/thanos/statefulset.go#L222-L259) anymore but it still needs to add an annotation storing the secret's hash to trigger a rollout on config change.

**Comment by mviswanathsai** [comment]
Right now, this is what I see: 
1. Use the secrets informer and get the secrets in the namespace by their names (which we have). 
2. Hash using the `createssetInputHash`
3. Then create the annotation and check against it each time.
4. Redeploy or not based on that.

We already have the large part of this structure laid out and the hash creation happens [here](https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/thanos/operator.go#L639).

What needs to be added:
We need to add the logic to fetch each of the config files from the range of: 
`ObjectStorageConfig` 
`TracingConfig`
`QueryConfig`
`AlertmanagerConfig`
`AlertRelabelConfig`
Also, of the above fields some have a special variant which is given more priority. That needs to be accounted for in the functions as well.
`ObjectStorageConfigFile`
`TracingConfigFile`
`AlertRelabelConfigFile`

**Comment by mviswanathsai** [comment]
Also, since we ensure that the Thanos sset exists [here](https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/thanos/operator.go#L505). Is it possible that we can just hash all the contents of the [config directory](https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/thanos/statefulset.go#L40) and move from there rather than fetching each of the secret?
Just thinking out loud here. I did not come across something like this in the code base, maybe its not a good approach. But would like to hear thoughts on this.

**Comment by simonpasquier** [comment]
Yes I don't see how we can avoid retrieving all referenced secret data to generate the hash and save it to a pod annotation though it's quite heavy to do.

I'd start by writing an e2e test demonstrating the issue. I'd also check with the Thanos project if there's been any change. I haven't monitored closely but maybe it's now able to live-reload some of the files?

**Comment by mviswanathsai** [comment]
I am unsure where to look, but I checked [here](https://github.com/thanos-io/thanos/blob/main/pkg/reloader/reloader.go). Looks like it is possible to watch directories and call reload the *Prometheus instances* but its not possible to reload the ruler. Am I getting this right?

**Comment by mviswanathsai** [comment]
IIUC, we would like to ensure that the Thanos sset is enqueued for reconciliation on updates the Secrets in its namespace. I can see event handlers for ConfigMaps but none for Secrets, would we need to add it?
